sql: crash with subqueries and insert returning

```
root@127.0.0.1:54397/defaultdb> create table a( b int, c int);
CREATE TABLE

Time: 3.649ms

root@127.0.0.1:54397/defaultdb> select * from [insert into a values (1,2) returning b];
  b
+---+
  1
(1 row)

Time: 3.647ms

root@127.0.0.1:54397/defaultdb> select * from a where exists (select * from [insert into a values (1,2) returning b]);
E181003 21:10:11.962494 567 sql/conn_executor.go:663  [n1,client=127.0.0.1:54399,user=root] a SQL panic has occurred while executing "SELECT * FROM a WHERE EXISTS (SELECT * FROM [INSERT INTO a VALUES (1, 2) RETURNING b])": invalid row length 0, expected 1
E181003 21:10:11.962629 567 util/log/crash_reporting.go:203  [n1,client=127.0.0.1:54399,user=root] a panic has occurred!
E181003 21:10:11.964711 567 util/log/crash_reporting.go:477  [n1,client=127.0.0.1:54399,user=root] Reported as error 748791229d4e456f8bffa766016715da
panic: invalid row length 0, expected 1 [recovered]
        panic: panic while executing 1 statements: SELECT * FROM _ WHERE EXISTS (SELECT * FROM [INSERT INTO _ VALUES (_, _) RETURNING _]); caused by invalid row length 0, expected 1

goroutine 567 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00165c000, 0x6cc5440, 0xc000eef040, 0x61b61c0, 0xc001329090)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:677 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00165c000, 0x6cc5440, 0xc000eef040)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:416 +0x61
panic(0x61b61c0, 0xc001329090)
        /usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowContainer).AddRow(0xc001325dd0, 0x6cc5500, 0xc000ff48d0, 0x8570008, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/row_container.go:280 +0x462
github.com/cockroachdb/cockroach/pkg/sql.(*spoolNode).startExec(0xc000ff4510, 0x6cc5500, 0xc000ff48d0, 0xc000f3b0e0, 0xc00165c430, 0x6cb6001, 0xc00169f440)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/spool.go:70 +0x232
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x65ee8fd, 0x5, 0x6cb6580, 0xc000ff4510, 0xc0000f5980, 0xc001796db0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:641 +0x91
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc0007950e0, 0x65ee8fd, 0x5, 0x6cb6580, 0xc000ff4510)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:127 +0x5d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0007950e0, 0x6cb6580, 0xc000ff4510, 0x65ee8fd, 0x5)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:536 +0x1c4
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0007950e0, 0x6cb6580, 0xc000ff4510, 0xc001797910, 0x400e1ff)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:94 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(0x6cc5500, 0xc000ff48d0, 0x6cb6580, 0xc000ff4510, 0x0, 0x67480d8, 0x0, 0x0, 0xc000ff4a80, 0xc000000001, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:58 +0x1b2
github.com/cockroachdb/cockroach/pkg/sql.startExec(0x6cc5500, 0xc000ff48d0, 0xc000f3b0e0, 0xc00165c430, 0x6cb6580, 0xc000ff4510, 0xc000765c00, 0x6cc5500)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:646 +0xf8
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc000377900, 0x6cc5500, 0xc000ff48d0, 0xc000389580, 0x61b5440)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:123 +0xd9
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000377900, 0x6cc5500, 0xc000ff48d0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:767 +0x56
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).StartSync(0xc0013e8e00, 0x6cc5500, 0xc000ff48d0, 0x6747930, 0xc000cd6460, 0x6c9fe80)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:602 +0x19a
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0007cf500, 0xc000dd0480, 0xc0015393b0, 0xc0017987a8, 0xc0007fd340, 0xc000f3ad20, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:255 +0x8c9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunSubqueries(0xc0007cf500, 0x6cc5500, 0xc001347770, 0xc00165c430, 0xc000ff4630, 0xc001195180, 0x1, 0x1, 0xc0007fd080, 0xc00165c101, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:664 +0x845
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00165c000, 0x6cc5500, 0xc001347770, 0xc00165c430, 0x3, 0xd9c4678, 0xc0015395f0, 0x1, 0x0, 0x0)
```

I think the bug here is that the `rowSourceToPlanNode` blindly reports the columns of the `planNode` it replaced, without taking into account that some of the columns might be projected out due to being omitted by the needed column infrastructure. I fixed that, and the panic went away. But another similar query is still crashing:

```
root@127.0.0.1:54675/defaultdb> select 1 from [insert into a values(1,2) returning b]  join [insert into a values(2,3) returning b] on true;
foo []
[]
foo []
[]
E181003 22:06:23.269408 497 sql/conn_executor.go:663  [n1,client=127.0.0.1:54677,user=root] a SQL panic has occurred while executing "SELECT 1 FROM [INSERT INTO a VALUES (1, 2) RETURNING b] JOIN [INSERT INTO a VALUES (2, 3) RETURNING b] ON true": decoding unset EncDatum
E181003 22:06:23.269538 497 util/log/crash_reporting.go:203  [n1,client=127.0.0.1:54677,user=root] a panic has occurred!
E181003 22:06:23.276757 497 util/log/crash_reporting.go:477  [n1,client=127.0.0.1:54677,user=root] Reported as error e994385a3183471d833898d0241b5229
panic: decoding unset EncDatum [recovered]
        panic: panic while executing 1 statements: SELECT _ FROM [INSERT INTO _ VALUES (_, _) RETURNING _] JOIN [INSERT INTO _ VALUES (_, _) RETURNING _] ON _; caused by decoding unset EncDatum

goroutine 497 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000f8e000, 0x6cf0000, 0xc000887900, 0x61d7fe0, 0x6cb3e50)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:677 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000f8e000, 0x6cf0000, 0xc000887900)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:416 +0x61
panic(0x61d7fe0, 0x6cb3e50)
        /usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*EncDatum).EnsureDecoded(0xc000f61ef0, 0xc00084d340, 0xc0004aa610, 0x1, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:210 +0x540
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*memRowContainer).AddRow(0xc0004aa510, 0x6cf00c0, 0xc0008b8330, 0xc000f61ef0, 0x1, 0x1, 0x639ce00, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/row_container.go:166 +0xc6
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*hashJoiner).build(0xc0004a9b00, 0x400ed68, 0x20, 0x639cee0, 0xc000331f01, 0xc001563f40)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/hashjoiner.go:341 +0x198
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*hashJoiner).Next(0xc0004a9b00, 0xb, 0x6cf00c0, 0xc0008b8330, 0x6cf00c0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/hashjoiner.go:243 +0x1ba
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x6cf00c0, 0xc0008b8330, 0x6cfa2c0, 0xc0004a9b00, 0x6cca800, 0xc000eb5180)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:170 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc0004a9b00, 0x6cf00c0, 0xc0008b8330, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:768 +0x96
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).StartSync(0xc000db68c0, 0x6cf00c0, 0xc0008b8330, 0x676f370, 0xc0008ac000, 0x6cca440)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:602 +0x19a
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000982100, 0xc0014a3320, 0xc000bfb680, 0xc00169e930, 0xc000bf82c0, 0xc000f8e4c8, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:255 +0x8c9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000982100, 0x6cf00c0, 0xc000f61260, 0xc000f8e4c8, 0xc0014a3320, 0xc000bfb680, 0x6ce0880, 0xc0011dc840, 0xc000bf82c0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:747 +0x24c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000f8e000, 0x6cf00c0, 0xc000f61260, 0xc000f8e430, 0x3, 0xb244818, 0xc000bfb710, 0x0, 0x0, 0x0)
```

--- a/pkg/sql/logictest/testdata/logic_test/statement_source
+++ b/pkg/sql/logictest/testdata/logic_test/statement_source
@@ -70,3 +70,23 @@ SELECT * FROM a ORDER BY b
 100  2
 -2   3
 7    8
+
+# Regression for #30936: ensure that wrapped planNodes with non-needed columns work ok
+
+statement ok
+CREATE TABLE b (a int, b int)
+
+query II
+SELECT * FROM b WHERE EXISTS (SELECT * FROM [INSERT INTO b VALUES (1,2) RETURNING a,b]);
+----
+1 2
+
+query I
+SELECT 1 FROM [INSERT INTO b VALUES(2,3) RETURNING b] JOIN [INSERT INTO b VALUES(4,5) RETURNING b] ON true;
+----
+1
+
+query III
+SELECT * FROM [INSERT INTO b VALUES(2,3) RETURNING b] JOIN [INSERT INTO b VALUES(4,5) RETURNING b, a] ON true;
+----
+3 5 4

--- a/pkg/sql/plan_node_to_row_source.go
+++ b/pkg/sql/plan_node_to_row_source.go
@@ -105,7 +105,23 @@ func (p *planNodeToRowSource) SetInput(ctx context.Context, input distsqlrun.Row
 	return walkPlan(ctx, p.node, planObserver{
 		replaceNode: func(ctx context.Context, nodeName string, plan planNode) (planNode, error) {
 			if plan == p.firstNotWrapped {
-				return makeRowSourceToPlanNode(input, p, planColumns(p.firstNotWrapped), p.firstNotWrapped), nil
+				cols := planColumns(p.firstNotWrapped)
+				nOmitted := 0
+				for i := range cols {
+					if cols[i].Omitted {
+						nOmitted++
+					}
+				}
+				if nOmitted > 0 {
+					newCols := make(sqlbase.ResultColumns, 0, len(cols)-nOmitted)
+					for i := range cols {
+						if !cols[i].Omitted {
+							newCols = append(newCols, cols[i])
+						}
+					}
+					cols = newCols
+				}
+				return makeRowSourceToPlanNode(input, p, cols, p.firstNotWrapped), nil
 			}
 			return nil, nil
 		},

